[kep] performance tests (#1640)
Kong/kubernetes-ingress-controller: keps/0007-kic-performance-test.md

---
title: KIC Performance Test Solution

> status: implementable

Reviewer: We need to start this out in a `provisional` status, as this will indicate that we're still making some core decisions about the proposal and nobody should be starting an implementation (save perhaps a spike or prototype to support the KEP) yet:

```suggestion
status: provisional
```

---

# Notes
For references see the dependencies related to this proprosal to check the progress of related efforts:
- solution 
- deployment and configuration
- code/repo
- configuration and execution
- profiling (code profiler)

> - histogram execution time and calculates percentiles

Reviewer: I'm not sure I understand this notes sections, perhaps we can add URL links?

Author: similar as https://perf.dashboard.istio.io/benchmarks/cpu_vs_conn/

Reviewer: Ok... still not sure I understand these notes HOWEVER they are just notes and I don't want to block on them, we can always update later.

Author: hmm... the execution time of request, API, which basically a metrics diagram, similar as profiling.

## Summary

>   We want a performance test solution that can collect controller metrics per release, which also provides profiling
> and histogram execution time. The solution is able to scale according to controller scalability. We mainly use kind as test cluster, but also support varied cloud provider cluster such EKS, AKS, GKE etc.

Reviewer: I have a slightly different idea of what our summary should be based on some desire to use language that makes things more generic (and less specific to our use cases):

```suggestion

Testing libraries and tools for reporting performance metrics of the Kong Kubernetes Ingress Controller (KIC) in any Kubernetes cluster environment.
```

Let me know your thoughts, and whether this wording resonates with you or not?

Author: A more generic description is : we need testing tools/framekwork such as jenkins, KIC, or curl finish testing, and test report could be publicly accessible downloadable.

Reviewer: Personally I think pointing out specific tools in this doc (especially this iteration) makes us _less_ generic. I would like to see a truly high level summary that states the basic thing we want to accomplish with as few details as necessary.

Author: updated accordingly.

Reviewer: I've read the new one, this reads well for me so let's go with this for now :+1:

[kic]:https://github.com/kong/kubernetes-ingress-controller
[ktf]:https://github.com/Kong/kubernetes-testing-framework
[jenkins]: https://www.jenkins.io/
[aws]: https://aws.amazon.com/
[kind]: https://kind.sigs.k8s.io/
[docker]: https://www.docker.com/ 

## Motivation

>   - provide an isolated and reproducible environment for performance testing
>   - continue integration and continue test
>   - Visualize testing configuration and results.
>   - System distribution

Reviewer: Generally the motivation section should be causal factors, whereas these all appear to be more like goals in that they're specific things we want to achieve.

I would like to suggest the following alternative for motivations:

```suggestion
- we currently have no profiles, documentation, or determined performance characteristics for the KIC in any Kubernetes or infrastructure environment
- we want libraries and tooling that can be run against a Kubernetes cluster and its underlying infrastructure to analyze the performance characteristics of the KIC in that environment
- we want our tooling to be open source so end-users can run a performance analysis themselves in their environment
```

Based on the history of this proposal that I'm aware of, I think that this covers some of the motivations that originally led to this project earlier in the year. Let me know your thoughts?

Author: it looks like the opposite way presenting the same thing, 
this PR - want to do A,B,C
You are suggesting: we do not have A, B, C

I do not have strong preference, would like to follow your thought.

Reviewer: Right, "we do not have A, B, C" provides the motivators. The "motivation" are the things that are currently true that have lead us to want to propose a change.

### Goals

>   - Able to run performance pipeline repeatedly
>   - Able to collect following performance metrics
>     - controller profiling
>     - controller memory/cpu consumption
>     - ingress config, deck sync, parser cost

Reviewer: I think we should point out "what" is "able to do X" so that our goals are clearly providing some direction towards the eventual implementation details. I also personally would like to point out exactly how we intend to implement this at a very high level:

```suggestion
- Develop [Golang](https://go.dev) libraries that are able to run a performance testing pipeline on a Kubernetes cluster and collect the following metrics:
  - controller memory/cpu consumption
  - ingress config, deck sync, parser cost
  - controller profiling (pprof data)
- Develop a CLI interface to the Golang libraries for general purpose usage
- Develop a container image to run the CLI in a container for automated deployments
- Develop a [Helm Chart](https://helm.sh) for the container image to easily deploy it to Kubernetes clusters
```

Author: Would say the Performance Test Solution is mainly for performance testing purpose, docker container, pipeline fits well, helm chart is good suggestion though needs think about it. 
Interms of the CLM, pipeline works as similar, and the configuration items are similar as command line parameters. Would take that as an enhancement.
